File: src/main/java/co/herod/adbwrapper/AdbBusManager.java
package co.herod.adbwrapper;

import co.herod.adbwrapper.model.AdbUiHierarchy;
import co.herod.adbwrapper.model.AdbUiNode;
import co.herod.adbwrapper.util.Utils;
import io.reactivex.Observable;
import org.jetbrains.annotations.NotNull;

/**
 * Created by matthewherod on 23/04/2017.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class AdbBusManager {

    @NotNull
    public static final BusSubject<String> ADB_BUS = new MainBusSubject();

    @NotNull
    public static final BusSubject<AdbUiHierarchy> ADB_UI_HIERARCHY_BUS = new AdbUiHierarchyBus();

    @NotNull
    public static final BusSubject<AdbUiNode> ADB_UI_NODE_BUS = new AdbUiNodeBus();

    private AdbBusManager() {
    }

    static Observable<String> throttledBus() {
        return ADB_BUS.concatMap(Utils::throttleOutput);
    }

    private static class MainBusSubject extends BusSubject<String> {

    }

    private static class AdbUiHierarchyBus extends BusSubject<AdbUiHierarchy> {

    }

    private static class AdbUiNodeBus extends BusSubject<AdbUiNode> {

    }
}

File: src/main/java/co/herod/adbwrapper/AdbBusManager.kt
package co.herod.adbwrapper

import co.herod.adbwrapper.model.AdbUiHierarchy
import co.herod.adbwrapper.model.AdbUiNode
import co.herod.adbwrapper.util.Utils
import io.reactivex.Observable

/**
 * Created by matthewherod on 23/04/2017.
 */
object AdbBusManager : IAdbBusManager {

    private val _ADB_BUS: BusSubject<String> = MainBusSubject()
    override fun getAdbBus(): BusSubject<String> = _ADB_BUS

    private val _ADB_UI_HIERARCHY_BUS: BusSubject<AdbUiHierarchy> = AdbUiHierarchyBus()
    override fun getAdbUiHierarchyBus(): BusSubject<AdbUiHierarchy> = _ADB_UI_HIERARCHY_BUS

    private val _ADB_UI_NODE_BUS: BusSubject<AdbUiNode> = AdbUiNodeBus()
    override fun getAdbUiNodeBus(): BusSubject<AdbUiNode> = _ADB_UI_NODE_BUS

    internal fun throttledBus(): Observable<String>? = getAdbBus().concatMap { Utils.throttleOutput(it) }

    private class MainBusSubject : BusSubject<String>()

    private class AdbUiHierarchyBus : BusSubject<AdbUiHierarchy>()

    private class AdbUiNodeBus : BusSubject<AdbUiNode>()
}
